src/System.Security.Claims/src/System/Security/Claims/Claim.cs

```diff
@@ -0,0 +1,550 @@
+            : this(type, value, valueType, issuer, originalIssuer, subject, null, null)
+        {
+        }
+
+        /// <summary>
+        /// This internal constructor was added as a performance boost when adding claims that are found in the NTToken.
+        /// We need to add a property value to distinguish DeviceClaims from UserClaims.
+        /// </summary>
+        /// <param name="propertyKey">This allows adding a property when adding a Claim.</param>
+        /// <param name="propertyValue">The value associcated with the property.</param>
+        internal Claim(string type, string value, string valueType, string issuer, string originalIssuer, ClaimsIdentity subject, string propertyKey, string propertyValue)
+        {
+            if (type == null)
+            {
+                throw new ArgumentNullException("type");
+            }
+
+            if (value == null)
+            {
+                throw new ArgumentNullException("value");
+            }
+
+            Contract.EndContractBlock();
+
+            _type = type;
+            _value = value;
+            _valueType = string.IsNullOrEmpty(valueType) ? ClaimValueTypes.String : valueType;
+            _issuer = string.IsNullOrEmpty(issuer) ? ClaimsIdentity.DefaultIssuer : issuer;
+            _originalIssuer = string.IsNullOrEmpty(originalIssuer) ? _issuer : originalIssuer;
+            _subject = subject;
+
+            if (propertyKey != null)
+            {
+                _properties[propertyKey] = propertyValue;
+            }
+        }
+
+        /// <summary>
+        /// Copy constructor for <see cref="Claim"/>
+        /// </summary>
+        /// <param name="other">the <see cref="Claim"/> to copy.</param>
+        /// <remarks><see cref="Claim.Subject"/>will be set to 'null'.</remarks>
+        /// <exception cref="ArgumentNullException">if 'other' is null.</exception>
+        protected Claim(Claim other)
+            : this(other, (other == null ? (ClaimsIdentity)null : other._subject))
+        {
+        }
+
+        /// <summary>
+        /// Copy constructor for <see cref="Claim"/>
+        /// </summary>
+        /// <param name="other">the <see cref="Claim"/> to copy.</param>
+        /// <param name="subject">the <see cref="ClaimsIdentity"/> to assign to <see cref="Claim.Subject"/>.</param>
+        /// <remarks><see cref="Claim.Subject"/>will be set to 'subject'.</remarks>
+        /// <exception cref="ArgumentNullException">if 'other' is null.</exception>
+        protected Claim(Claim other, ClaimsIdentity subject)
+        {
+            if (other == null)
+                throw new ArgumentNullException("other");
+
+            _issuer = other._issuer;
+            _originalIssuer = other._originalIssuer;
+            _subject = subject;
+            _type = other._type;
+            _value = other._value;
+            _valueType = other._valueType;
+            foreach (var key in other._properties.Keys)
```

Comment:
Enumerating `other._properties.Keys` is going to allocate an instance of `KeysCollection<TKey>`. This can be avoided by just looping over the dictionary directly, e.g.

```c#
foreach (var pair in other._properties)
{
    _properties.Add(pair.Key, pair.Value);
}
```